CIS crashes with "invalid memory address or nil pointer dereference" as soon as CIS-args "namespace-label" is used!

#### Setup Details
 CIS Version : 2.12.1      
 Build: f5networks/k8s-bigip-ctlr:latest       
 BIGIP Version: Big IP 15.1.8.1.0.32.3      
 AS3 Version: 3.34      
 Agent Mode: AS3       
 Orchestration: K8S/OSCP v4.9.xx, Kubernetes v1.24.0+3882f8f  
 Orchestration Version:       
 Pool Mode: Cluster


#### Description  
I use the CIS in legacy mode (OpenShift Routes).
The CIS crashes as soon as I switch to CIS-option "--namespace-label=F5_env=foo".
When I switch back to my former multiple CIS-options "--namespace=foo1, ""--namespace=foo2" etc. the CIS does not crash.


#### Steps To Reproduce
1) Remove CIS-options "--namespace=foo1, ""--namespace=foo2" etc.
2) Add CIS-option "--namespace-label=F5_env=foo".
3) Start CIS

#### Expected Result 
CIS should start normally and should watch Namespaces which use the specified namespace-label.

#### Actual Result  
CIS crashes at starting.

#### Diagnostic Information
```buildoutcfg
[...]
2023/05/23 13:10:23 [DEBUG] [CORE] Finished syncing namespace xyz (100.315733ms)
2023/05/23 13:10:23 [DEBUG] [CORE] Creating new app informer
2023/05/23 13:10:23 [DEBUG] [CCCL] ConfigWriter (0xc0002b7b60) writing section name vxlan-fdb
2023/05/23 13:10:23 [INFO] [CORE] Not watching Ingress resources.
E0523 13:10:23.792194       1 runtime.go:78] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 215 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic({0x167fae0?, 0x2697750})
	/go/src/github.com/F5Networks/k8s-bigip-ctlr/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:74 +0x99
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0x80?})
	/go/src/github.com/F5Networks/k8s-bigip-ctlr/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:48 +0x75
panic({0x167fae0, 0x2697750})
	/usr/local/go/src/runtime/panic.go:884 +0x212
github.com/F5Networks/k8s-bigip-ctlr/v2/pkg/appmanager.(*Manager).setupNodeProcessing(0xc000144000)
	/go/src/github.com/F5Networks/k8s-bigip-ctlr/pkg/appmanager/appManager.go:3543 +0x99
github.com/F5Networks/k8s-bigip-ctlr/v2/pkg/appmanager.(*Manager).newAppInformer.func21({0xc00099fe38?, 0x1?})
	/go/src/github.com/F5Networks/k8s-bigip-ctlr/pkg/appmanager/appManager.go:858 +0x1d
k8s.io/client-go/tools/cache.ResourceEventHandlerFuncs.OnAdd(...)
	/go/src/github.com/F5Networks/k8s-bigip-ctlr/vendor/k8s.io/client-go/tools/cache/controller.go:231
k8s.io/client-go/tools/cache.(*processorListener).run.func1()
	/go/src/github.com/F5Networks/k8s-bigip-ctlr/vendor/k8s.io/client-go/tools/cache/shared_informer.go:777 +0x134
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0x26ea2a0?)
	/go/src/github.com/F5Networks/k8s-bigip-ctlr/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:155 +0x3e
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc00042e738?, {0x1ae3220, 0xc000520060}, 0x1, 0xc000224120)
	/go/src/github.com/F5Networks/k8s-bigip-ctlr/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:156 +0xb6
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0x10?, 0x3b9aca00, 0x0, 0x31?, 0xc00042e788?)
	/go/src/github.com/F5Networks/k8s-bigip-ctlr/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0x89
k8s.io/apimachinery/pkg/util/wait.Until(...)
	/go/src/github.com/F5Networks/k8s-bigip-ctlr/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:90
k8s.io/client-go/tools/cache.(*processorListener).run(0xc0000c4300?)
	/go/src/github.com/F5Networks/k8s-bigip-ctlr/vendor/k8s.io/client-go/tools/cache/shared_informer.go:771 +0x6b
k8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1()
	/go/src/github.com/F5Networks/k8s-bigip-ctlr/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:73 +0x5a
created by k8s.io/apimachinery/pkg/util/wait.(*Group).Start
	/go/src/github.com/F5Networks/k8s-bigip-ctlr/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:71 +0x85
2023/05/23 13:10:23 [INFO] [CORE] Watching ConfigMap resources.
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x90 pc=0x1420e59]
[...]


```  

#### Observations (if any)

--- a/cmd/k8s-bigip-ctlr/main.go
+++ b/cmd/k8s-bigip-ctlr/main.go
@@ -751,6 +751,10 @@ func GetNamespaces(appMgr *appmanager.Manager) {
 func setupWatchers(appMgr *appmanager.Manager, resyncPeriod time.Duration) {
 	label := resource.DefaultConfigMapLabel
 
+	err := appMgr.AddNodeInformer(resyncPeriod)
+	if nil != err {
+		log.Warningf("[INIT] Failed to add node informer for the controller:%v", err)
+	}
 	if len(*namespaceLabel) == 0 {
 		// For periodic monitoring
 		// Non monitoring namespaces will not be processed

--- a/docs/RELEASE-NOTES.rst
+++ b/docs/RELEASE-NOTES.rst
@@ -15,6 +15,7 @@ Bug Fixes
 * Exclude the removal of static ARP entries for Flannel CNI during CIS restart
 * `Issue 2800 <https://github.com/F5Networks/k8s-bigip-ctlr/issues/2800>`_: Fix monitor not creating for VS CRD when send string is missing
 * `Issue 2867 <https://github.com/F5Networks/k8s-bigip-ctlr/issues/2867>`_: Ignore virtualServerName if hostGroup configured
+* `Issue 2898 <https://github.com/F5Networks/k8s-bigip-ctlr/issues/2898>`_: Fix for CIS crash with namespace-label parameter
 
 2.13.0
 -------------

--- a/pkg/appmanager/appManager.go
+++ b/pkg/appmanager/appManager.go
@@ -97,7 +97,7 @@ type Manager struct {
 	// App informer support
 	vsQueue      workqueue.RateLimitingInterface
 	appInformers map[string]*appInformer
-	as3Informer  *appInformer
+	nodeInformer *nodeInformer
 	// Namespace informer support (namespace labels)
 	nsQueue    workqueue.RateLimitingInterface
 	nsInformer cache.SharedIndexInformer
@@ -484,6 +484,38 @@ func (appMgr *Manager) removeNamespaceLocked(namespace string) error {
 	return nil
 }
 
+// AddNodeInformer to watch the node udpates
+func (appMgr *Manager) AddNodeInformer(resyncPeriod time.Duration) error {
+	appMgr.informersMutex.Lock()
+	defer appMgr.informersMutex.Unlock()
+
+	nodeOptions := func(options *metav1.ListOptions) {
+		options.LabelSelector = appMgr.nodeLabelSelector
+	}
+	appMgr.nodeInformer = &nodeInformer{
+		stopCh: make(chan struct{}),
+		nodeInformer: cache.NewSharedIndexInformer(
+			cache.NewFilteredListWatchFromClient(
+				appMgr.restClientv1,
+				"nodes",
+				"",
+				nodeOptions,
+			),
+			&v1.Node{},
+			resyncPeriod,
+			cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc},
+		),
+	}
+	appMgr.nodeInformer.nodeInformer.AddEventHandler(
+		&cache.ResourceEventHandlerFuncs{
+			AddFunc:    func(obj interface{}) { appMgr.setupNodeProcessing() },
+			UpdateFunc: func(obj, cur interface{}) { appMgr.setupNodeProcessing() },
+			DeleteFunc: func(obj interface{}) { appMgr.setupNodeProcessing() },
+		},
+	)
+	return nil
+}
+
 // AddNamespaceLabelInformer spins an informer to watch all namespaces with matching label
 func (appMgr *Manager) AddNamespaceLabelInformer(
 	labelSelector labels.Selector,
@@ -701,6 +733,11 @@ type serviceQueueKey struct {
 	Operation    string
 }
 
+type nodeInformer struct {
+	nodeInformer cache.SharedIndexInformer
+	stopCh       chan struct{}
+}
+
 type appInformer struct {
 	namespace        string
 	cfgMapInformer   cache.SharedIndexInformer
@@ -711,7 +748,6 @@ type appInformer struct {
 	secretInformer   cache.SharedIndexInformer
 	ingClassInformer cache.SharedIndexInformer
 	podInformer      cache.SharedIndexInformer
-	nodeInformer     cache.SharedIndexInformer
 	stopCh           chan struct{}
 }
 
@@ -723,9 +759,6 @@ func (appMgr *Manager) newAppInformer(
 	log.Debugf("[CORE] Creating new app informer")
 	everything := func(options *metav1.ListOptions) {
 		options.LabelSelector = ""
-	}
-	nodeOptions := func(options *metav1.ListOptions) {
-		options.LabelSelector = appMgr.nodeLabelSelector
 	}
 	appInf := appInformer{
 		namespace: namespace,
@@ -767,18 +800,6 @@ func (appMgr *Manager) newAppInformer(
 			cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc},
 		)
 	}
-
-	appInf.nodeInformer = cache.NewSharedIndexInformer(
-		cache.NewFilteredListWatchFromClient(
-			appMgr.restClientv1,
-			"nodes",
-			"",
-			nodeOptions,
-		),
-		&v1.Node{},
-		resyncPeriod,
-		cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc},
-	)
 
 	//For nodeport mode, disable ep informer
 	if appMgr.poolMemberType != NodePort {
@@ -925,16 +946,6 @@ func (appMgr *Manager) newAppInformer(
 				AddFunc:    func(obj interface{}) { appMgr.enqueuePod(obj, OprTypeCreate) },
 				UpdateFunc: func(obj, cur interface{}) { appMgr.enqueuePod(cur, OprTypeUpdate) },
 				DeleteFunc: func(obj interface{}) { appMgr.enqueuePod(obj, OprTypeDelete) },
-			},
-		)
-	}
-
-	if appInf.nodeInformer != nil {
-		appInf.nodeInformer.AddEventHandler(
-			&cache.ResourceEventHandlerFuncs{
-				AddFunc:    func(obj interface{}) { appMgr.setupNodeProcessing() },
-				UpdateFunc: func(obj, cur interface{}) { appMgr.setupNodeProcessing() },
-				DeleteFunc: func(obj interface{}) { appMgr.setupNodeProcessing() },
 			},
 		)
 	}
@@ -1118,9 +1129,6 @@ func (appInf *appInformer) start() {
 	if nil != appInf.podInformer {
 		go appInf.podInformer.Run(appInf.stopCh)
 	}
-	if nil != appInf.nodeInformer {
-		go appInf.nodeInformer.Run(appInf.stopCh)
-	}
 }
 
 func (appInf *appInformer) waitForCacheSync() {
@@ -1143,9 +1151,6 @@ func (appInf *appInformer) waitForCacheSync() {
 	}
 	if nil != appInf.cfgMapInformer {
 		cacheSyncs = append(cacheSyncs, appInf.cfgMapInformer.HasSynced)
-	}
-	if nil != appInf.nodeInformer {
-		cacheSyncs = append(cacheSyncs, appInf.nodeInformer.HasSynced)
 	}
 	if nil != appInf.ingClassInformer {
 		cacheSyncs = append(cacheSyncs, appInf.ingClassInformer.HasSynced)
@@ -1180,13 +1185,14 @@ func (appMgr *Manager) runImpl(stopCh <-chan struct{}) {
 	defer utilruntime.HandleCrash()
 	defer appMgr.vsQueue.ShutDown()
 	defer appMgr.nsQueue.ShutDown()
-
+	// start and sync node informer.
+	appMgr.startAndSyncNodeInformer()
 	if nil != appMgr.nsInformer {
 		// Using one worker for namespace label changes.
 		appMgr.startAndSyncNamespaceInformer(stopCh)
 		go wait.Until(appMgr.namespaceWorker, time.Second, stopCh)
 	}
-
+	// start and sync App informer
 	appMgr.startAndSyncAppInformers()
 
 	// Using only one virtual server worker currently.
@@ -1194,6 +1200,14 @@ func (appMgr *Manager) runImpl(stopCh <-chan struct{}) {
 
 	<-stopCh
 	appMgr.stopAppInformers()
+	close(appMgr.nodeInformer.stopCh)
+}
+
+func (appMgr *Manager) startAndSyncNodeInformer() {
+	appMgr.informersMutex.Lock()
+	defer appMgr.informersMutex.Unlock()
+	go appMgr.nodeInformer.nodeInformer.Run(appMgr.nodeInformer.stopCh)
+	cache.WaitForCacheSync(appMgr.nodeInformer.stopCh, appMgr.nodeInformer.nodeInformer.HasSynced)
 }
 
 func (appMgr *Manager) startAndSyncNamespaceInformer(stopCh <-chan struct{}) {
@@ -1214,9 +1228,6 @@ func (appMgr *Manager) startAppInformersLocked() {
 	for _, appInf := range appMgr.appInformers {
 		appInf.start()
 	}
-	if nil != appMgr.as3Informer {
-		appMgr.as3Informer.start()
-	}
 }
 
 func (appMgr *Manager) waitForCacheSync() {
@@ -1229,9 +1240,6 @@ func (appMgr *Manager) waitForCacheSyncLocked() {
 	for _, appInf := range appMgr.appInformers {
 		appInf.waitForCacheSync()
 	}
-	if nil != appMgr.as3Informer {
-		appMgr.as3Informer.waitForCacheSync()
-	}
 }
 
 func (appMgr *Manager) stopAppInformers() {
@@ -1239,9 +1247,6 @@ func (appMgr *Manager) stopAppInformers() {
 	defer appMgr.informersMutex.Unlock()
 	for _, appInf := range appMgr.appInformers {
 		appInf.stopInformers()
-	}
-	if nil != appMgr.as3Informer {
-		appMgr.as3Informer.stopInformers()
 	}
 }
 
@@ -3651,17 +3656,7 @@ func (appMgr *Manager) setOtherSDNType() {
 func (appMgr *Manager) setupNodeProcessing() error {
 	// Register appMgr to watch for node updates to keep track of watched nodes
 	//when there is update from node informer get list of nodes from nodeinformer cache
-	ns := ""
-	if appMgr.watchingAllNamespacesLocked() {
-		ns = ""
-	} else {
-		for _, k := range appMgr.GetAllWatchedNamespaces() {
-			ns = k
-			break
-		}
-	}
-	appInf, _ := appMgr.getNamespaceInformer(ns)
-	nodes := appInf.nodeInformer.GetIndexer().List()
+	nodes := appMgr.nodeInformer.nodeInformer.GetIndexer().List()
 	var nodeslist []v1.Node
 	for _, obj := range nodes {
 		node := obj.(*v1.Node)

--- a/pkg/appmanager/appManager_test.go
+++ b/pkg/appmanager/appManager_test.go
@@ -406,8 +406,7 @@ func (m *mockAppManager) addNamespace(ns *v1.Namespace) bool {
 }
 
 func (m *mockAppManager) addNode(node *v1.Node, ns string) {
-	appInf, _ := m.appMgr.getNamespaceInformer(ns)
-	appInf.nodeInformer.GetStore().Add(node)
+	m.appMgr.nodeInformer.nodeInformer.GetStore().Add(node)
 	m.appMgr.setupNodeProcessing()
 }
 

--- a/pkg/appmanager/ingress_test.go
+++ b/pkg/appmanager/ingress_test.go
@@ -1039,6 +1039,8 @@ var _ = Describe("V1 Ingress Tests", func() {
 			Expect(err).To(BeNil())
 			mockMgr.appMgr.useNodeInternal = true
 			mockMgr.appMgr.WatchedNS = WatchedNamespaces{Namespaces: []string{namespace}}
+			_ = mockMgr.appMgr.AddNodeInformer(0)
+			mockMgr.appMgr.startAndSyncNodeInformer()
 			// Ingress first
 			ingressConfig := netv1.IngressSpec{
 				IngressClassName: &IngressClassName,
@@ -1067,8 +1069,7 @@ var _ = Describe("V1 Ingress Tests", func() {
 			node := test.NewNode("node1", "1", false,
 				[]v1.NodeAddress{{Type: "InternalIP", Address: "127.0.0.1"}}, []v1.Taint{})
 			mockMgr.addNode(node, namespace)
-			appInf, _ := mockMgr.appMgr.getNamespaceInformer(namespace)
-			Expect(len(appInf.nodeInformer.GetIndexer().List())).To(Equal(1))
+			Expect(len(mockMgr.appMgr.nodeInformer.nodeInformer.GetIndexer().List())).To(Equal(1))
 
 			// Create the services
 			fooSvc := test.NewService("foo", "1", namespace, "NodePort",

